test/no_implicit_any_test.ts

```diff
@@ -0,0 +1,142 @@
+    jasmine.addMatchers(SourceFileComparer);
+
+    const relativeOutputPath = './ts_upgrade';
+    const inputConfigPath = './test/test_files/tsconfig.json';
+
+    project = new Project({
+      tsConfigFilePath: inputConfigPath,
+      addFilesFromTsConfig: false,
+      compilerOptions: {
+        noImplicitAny: true,
+      },
+    });
+    errorDetector = new ProdErrorDetector();
+    emitter = new OutOfPlaceEmitter(relativeOutputPath);
+    manipulator = new NoImplicitAnyManipulator(errorDetector);
+  });
+
+  const testFiles = [
+    {
+      description: 'fixes basic cases of assignment and function calls',
+      inputFilePath:
+        './test/test_files/no_implicit_any/basic_assignment_and_functions.ts',
+      actualOutputFilePath:
+        './test/test_files/no_implicit_any/ts_upgrade/basic_assignment_and_functions.ts',
+      expectedOutputFilePath:
+        './test/test_files/golden/no_implicit_any/basic_assignment_and_functions.ts',
+    },
+    {
+      description:
+        'fixes declarations that require resolving in specific order',
+      inputFilePath:
+        './test/test_files/no_implicit_any/topologically_dependent.ts',
+      actualOutputFilePath:
+        './test/test_files/no_implicit_any/ts_upgrade/topologically_dependent.ts',
+      expectedOutputFilePath:
+        './test/test_files/golden/no_implicit_any/topologically_dependent.ts',
+    },
+  ];
+
+  for (let test of testFiles) {
+    it(test.description, () => {
+      const input = project.addSourceFileAtPath(test.inputFilePath);
+      project.resolveSourceFileDependencies();
+
+      new Runner(
+        /* args*/ undefined,
+        project,
+        /* parser */ undefined,
+        errorDetector,
+        [manipulator],
+        emitter
+      ).run();
+
+      const expectedOutput = project.addSourceFileAtPath(
+        test.expectedOutputFilePath
+      );
+      const actualOutput = project.addSourceFileAtPath(
+        test.actualOutputFilePath
+      );
+
+      expect(actualOutput).toHaveSameASTAs(expectedOutput);
+
+      project.removeSourceFile(input);
+      project.removeSourceFile(actualOutput);
+      project.removeSourceFile(expectedOutput);
+    });
+  }
+
+  it('topologically sorts graph', () => {
+    const basicGraph = {
+      vertices: new Set(['x', 'y', 'z', 'w']),
+      edges: new Map([
+        ['x', new Set(['y', 'z'])],
+        ['y', new Set(['z', 'w'])],
+      ]),
+      sorted: [
+        ['x', 'y', 'z', 'w'],
+        ['x', 'y', 'w', 'z'],
+      ],
+    };
+
+    expect(
+      basicGraph.sorted.some(expectedOut =>
+        _.isEqual(
+          expectedOut,
+          manipulator.topoSort(basicGraph.vertices, basicGraph.edges)
+        )
+      )
+    ).toBeTrue();
+
+    const cycleGraph = {
+      vertices: new Set(['x', 'y', 'z', 'w']),
+      edges: new Map([
+        ['x', new Set(['y'])],
+        ['y', new Set(['z'])],
+        ['z', new Set(['w'])],
+        ['w', new Set(['y'])],
+      ]),
+      sorted: ['x'],
```

Comment:
I ended up reimplementing the toposort so that instead of erroring out for cyclical dependencies, it returns an order of declarations that minimizes dependency violations. I also changed the way we're calculating the dependency types so that even in the case of cyclical dependencies, it should be able to add the right types. For the example you gave, the tool would assign type `string` to all 3 variables: y, z, w.